loki/transformations/pragma_model.py

```diff
@@ -0,0 +1,346 @@
+        if content:
+            return Pragma(keyword='omp', content=f'target update{content}')
+        return self.default_retval()
+
+    def pmap_unstructured_data(self, pragma, parameters, **kwargs):
+        content = ''
+        if param_in := parameters.get('in'):
+            content += f' map(to: {param_in})'
+        if param_create := parameters.get('create'):
+            content += f' map(alloc: {param_create})'
+        if content:
+            return Pragma(keyword='omp', content=f'target enter data{content}')
+        return self.default_retval()
+
+    def pmap_end_unstructured_data(self, pragma, parameters, **kwargs):
+        content = ''
+        if params_out := parameters.get('out'):
+            content += f' map(from: {params_out})'
+        if params_delete := parameters.get('delete'):
+            content += f' map(delete: {params_delete})'
+        if content:
+            return Pragma(keyword='omp', content=f'target exit data{content}')
+        return self.default_retval()
+
+    def pmap_structured_data(self, pragma, parameters, **kwargs):
+        content = ''
+        if params_in := parameters.get('in'):
+            content += f' map(to: {params_in})'
+        if params_inout := parameters.get('inout'):
+            content += f' map(tofrom: {params_inout})'
+        if params_out := parameters.get('out'):
+            content += f' map(from: {params_out})'
+        if params_create := parameters.get('create'):
+            content += f' map(alloc: {params_create})'
+        if content:
+            return Pragma(keyword='omp', content=f'target data{content}')
+        return self.default_retval()
+
+    def pmap_end_structured_data(self, pragma, parameters, **kwargs):
+        return Pragma(keyword='omp', content='end target data')
+
+    def pmap_routine(self, pragma, parameters, **kwargs):
+        if 'seq' in parameters:
+            return Pragma(keyword='omp', content='declare target')
+        return self.default_retval()
+
+    def pmap_loop(self, pragma, parameters, **kwargs):
+        if 'vector' in parameters:
+            # TODO: private and reduction clause?
+            content = 'parallel do'
+            return Pragma(keyword='omp', content=content)
+        if 'gang' in parameters:
+            # TODO: private clause?
+            vlength_param = parameters.get('vlength')
+            vlength = f' thread_limit({vlength_param})' if vlength_param else ''
+            content = f'target teams distribute{vlength}'
+            return Pragma(keyword='omp', content=content)
+        return self.default_retval()
+
+    def pmap_end_loop(self, pragma, parameters, **kwargs):
+        if 'vector' in parameters:
+            return Pragma(keyword='omp', content='end parallel do')
+        if 'gang' in parameters:
+            return Pragma(keyword='omp', content='end target teams distribute')
+        return self.default_retval()
+
+
+class OpenMPPragmaMapper(GenericPragmaMapper):
```

Comment:
[optional] Should we call this OpenMPThreadingPragmaMapper to explicitly differentiate from offloading?